### ISSUE TITLE
add a command to scale a dimension expression

### ISSUE THREAD

**Issue by u-fischer** [latex3/latex3#727]
Currently adding a scale factor to a dimension expression is rather difficult. None of the "natural" syntax really works. 

~~~~
\documentclass{article}


\begin{document}
\ExplSyntaxOn
%\dim_set:Nn \l_tmpa_dim {2*\dim_max:nn {10pt}{15pt}} % errors
\dim_set:Nn \l_tmpa_dim {\dim_max:nn {10pt}{15pt}*2}  % works
\dim_show:N\l_tmpa_dim
\dim_set:Nn \l_tmpa_dim {\dim_max:nn {10pt}{15pt}*0.5} %multiply with 0
\dim_show:N\l_tmpa_dim

% wish:
%\dim_set:Nn \l_tmpa_dim {\dim_scale:nn {0.5}{\dim_max:nn {10pt}{15pt}} }
\ExplSyntaxOff
\end{document}
~~~~

**Comment by car222222** [comment]
What happens when adding parentheses? :

\dim_set:Nn \l_tmpa_dim {2* ( \dim_max:nn {10pt}{15pt} ) }

This looks like a bug:

\dim_set:Nn \l_tmpa_dim {\dim_max:nn {10pt}{15pt}*0.5}  % multiplies by 0

What happens with    * ( 0.5 )    ?

**Comment by wspr** [comment]
I think this does it:
```
\cs_new:Npn \dim_scale:nn #1#2
  {
    #1 \__dim_eval:w #2 \__dim_eval_end:
  }
```
I agree would be a good addition.

**Comment by josephwright** [comment]
I was going to add the suggestion, but to fit with everything else we have, you'd want the scale factor to be an fp expr. At that point, isn't just
```latex
\fp_to_dim:n{2 * \dim_max:nn {10pt}{15pt}}
```
reasonable?

**Comment by davidcarlisle** [comment]
using fp comes at a price though, other `dim_` functions take e-tex dim expressions so quick multiplicaton by a `<factor>` which covers the main use cases, I suspect

```
\RequirePackage{l3benchmark}
\ExplSyntaxOn

\benchmark:n{\dim_set:Nn\l_tmpa_dim {\fp_to_dim:n{2 * 10pt}}}
\benchmark:n{\dim_set:Nn\l_tmpa_dim {2  \dimexpr 10pt\relax}}

\ExplSyntaxOff

\stop
```

produces

```
8.06e-5 seconds (407 ops)
4.82e-7 seconds (2.48 ops)
```

but I don't think we have an L3 `\dimexpr` ?   `\dim_use_without_the:n`

**Comment by josephwright** [comment]
Well yes, that's of course an issue. The reason I thought of the `fp` construct was that a putative `\dim_scale:nn` is only really consistent with everything else if the first arg. is an fp expr. At that point, using `\fp_eval:n {#1} \__dim_eval:w #2 \__dim_eval_end:` seems pretty similar in performance to simply recommending `\fp_to_dim:n`.

**Comment by davidcarlisle** [comment]
yes, I noticed afterwards that Will's suggestion above is similar to mine. We do document dim expr (which uses factor rather than fp) so I don't think the consistency argument for fp is too strong here, especially if we offer `\dim_use_without_the:n` rather than a `\dim_scale` command with two arguments, so that the multiplication just comes as a consequence of dim expr syntax.

I agree if you are going to make the factor using `\fp_eval:n ` then there is no real gain and your `fp_to_dim:n` suggestion is fine.

**Comment by davidcarlisle** [comment]
> Where's that?

all dim functions accept a `<dim expr>` and somewhere we must say what a dim expr is, which is basically the e-tex primitive syntax, so you can use `2\l_tmp_dim` in any such expression  using `<factor>` multiplication, so  `2\dim_internal_dim:n{10pt}` seems natural to me,

**Comment by u-fischer** [comment]
> and somewhere we must say what a dim expr is

I can't find a suitable definition in interface3.pdf. 

> I was going to add the suggestion, but to fit with everything else we have, you'd want the scale factor to be an fp expr. At that point, isn't just
> 
> ```
> \fp_to_dim:n{2 * \dim_max:nn {10pt}{15pt}}
> ```
> 
> reasonable?

While it works, it didn't came up when I asked the question, so I would say it is not a natural idea to check the fp-module. People (at least me ;-)) look into the dim module to find dim-calculations. 

Also I think a fast version that skips the `\fp_eval:n`  wouldn't be bad. You could always use `\dim_scale:nn{\fp_eval:n{...}}{dim expr}` too, couldn't you?
